src/utils/tickets.rs

```diff
@@ -0,0 +1,110 @@
+use crate::response_code::{Error, Result, WrapperErrorKind};
+use crate::tss2_esys::{TPM2B_DIGEST, TPMT_TK_HASHCHECK, TPMT_TK_VERIFIED};
+use crate::utils::tags::StructureTag;
+use crate::utils::Hierarchy;
+use log::error;
+use std::convert::{TryFrom, TryInto};
+
+/// Macro used for implementing try_from
+/// TssTicketType -> TicketType
+/// TicketType -> TssTicketType
+macro_rules! impl_ticket_try_froms {
+    ($ticket_type:ident, $tss_ticket_type:ident) => {
+        impl TryFrom<$ticket_type> for $tss_ticket_type {
+            type Error = Error;
+            fn try_from(ticket: $ticket_type) -> Result<Self> {
+                let digest = ticket.digest;
+                if digest.len() > 64 {
+                    return Err(Error::local_error(WrapperErrorKind::WrongParamSize));
+                }
+                let mut buffer = [0; 64];
+                buffer[..digest.len()].clone_from_slice(&digest[..digest.len()]);
+                Ok($tss_ticket_type {
+                    tag: <$ticket_type>::TAG.into(),
+                    hierarchy: ticket.hierarchy.rh(),
+                    digest: TPM2B_DIGEST {
+                        size: digest.len().try_into().unwrap(), // should not fail based on the checks done above
+                        buffer,
+                    },
+                })
+            }
+        }
+
+        impl TryFrom<$tss_ticket_type> for $ticket_type {
+            type Error = Error;
+
+            fn try_from(tss_ticket: $tss_ticket_type) -> Result<Self> {
+                match StructureTag::try_from(tss_ticket.tag) {
+                    Ok(val) => {
+                        if val != <$ticket_type>::TAG {
+                            return Err(Error::local_error(WrapperErrorKind::InconsistentParams));
+                        }
+                    }
+                    Err(why) => {
+                        error!("Failed to parsed tag: {}", why);
+                        return Err(Error::local_error(WrapperErrorKind::InvalidParam));
+                    }
+                };
+
+                let len = tss_ticket.digest.size.into();
+                let mut digest = tss_ticket.digest.buffer.to_vec();
+                digest.truncate(len);
```

Comment:
Reckon it should be an error if `len` is longer than the buffer size? It `truncate` won't fail in any way, just asking from a programming experience point of view

Reply:
Fixed